src/Service/IndexData.php

```diff
@@ -0,0 +1,272 @@
+
+/**
+ * Class to contain configuration data for a single index suffix
+ */
+class IndexData
+{
+
+    use Injectable;
+    use Extensible;
+
+    public const string CONTEXT_KEY = 'context';
+    public const string CONTEXT_KEY_DEFAULT = 'default';
+
+    public function __construct(private array $data, private string $suffix)
+    {
+    }
+
+    /**
+     * Index contexts
+     *
+     * @var array
+     */
+    public array $contexts = [];
+
+    public function getData(): array
+    {
+        return $this->data;
+    }
+
+    public function getSuffix(): string
+    {
+        return $this->suffix;
+    }
+
+    public function getClassData(): array
+    {
+        return $this->data['includeClasses'] ?? [];
+    }
+
+    public function getClassConfig(string $class): ?array
+    {
+        $classData = $this->getClassData();
+
+        if (!array_key_exists($class, $classData)) {
+            return null;
+        }
+
+        $spec = $classData[$class];
+
+        if (!is_array($spec)) {
+            return null;
+        }
+
+        return $classData[$class];
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getClasses(): array
+    {
+        $classes = $this->getClassData() ?? [];
+        $result = [];
+
+        foreach ($classes as $className => $spec) {
+            if ($spec === false) {
+                continue;
+            }
+
+            $result[] = $className;
+        }
+
+        return $result;
+    }
+
+    public function getContextKey(): string
+    {
+        $key = static::CONTEXT_KEY_DEFAULT;
+
+        if (!array_key_exists(static::CONTEXT_KEY, $this->data)) {
+            return $key;
+        }
+
+        return $this->data[static::CONTEXT_KEY];
+    }
+
+    public function withIndexContext(callable $callback): void
+    {
+        $contextKey = $this->getContextKey();
+        $contexts = $this->contexts;
+
+        if (!array_key_exists($contextKey, $contexts)) {
+            throw new InvalidArgumentException(sprintf('No context configured for key: "%s"', $contextKey));
+        }
+
+        $context = $contexts[$contextKey];
+
+        $wrappers = array_map(
+            function (IndexDataContextProvider $provider) {
+                return $provider->getContext();
+            },
+            array_values($context)
+        );
+
+        $next = function () use ($callback): mixed {
+            return $callback($this);
+        };
+
+        foreach (array_reverse($wrappers) as $wrapper) {
+            $next = function () use ($wrapper, $next): mixed {
+                return $wrapper($next, $this);
+            };
+        }
+
+        $next();
+    }
+
+    public function getFields(): array
+    {
+        $fields = [];
+
+        $defaultFields = [
+            // Default fields that relate to our DataObjects
+            IndexConfiguration::singleton()->getSourceClassField(),
+            DataObjectDocument::config()->get('base_class_field'),
```

Comment:
@chrispenny I'm thinking that this kinda defeats the purpose of the DocumentInterface if we're hard coding config from one type of Document. Wondering if this should move to config or similar so different objects could be registered for an index?

Reply:
Added a todo to clean this up at a later date